### PR TITLE
Tutorial on ONNX [part 1 of 3]

### PR DESCRIPTION
**Pull request by ThomasDelteil**
Tutorial on Inference using ONNX models

### CODE REVIEW

File: docs/tutorials/onnx/inference_on_onnx_model.md [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+This is a helper function to run a batch of data through the net and collate the outputs
+
+
+```python
+def run_batch(net, data):
+    results = []
+    for batch in data:
+        outputs = net(batch)
+        for output in outputs.asnumpy(): results.append(output)
```
Comment (safrooze) [inline docs/tutorials/onnx/inference_on_onnx_model.md]: `results.extend([o for o in outputs.asnumpy()]`

File: docs/tutorials/onnx/inference_on_onnx_model.md [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    return np.array(results)
```
Comment (safrooze) [inline docs/tutorials/onnx/inference_on_onnx_model.md]: Can't understand what you're doing here. First you convert np.ndarray to a list, then convert it back to ndarray. What's changing?

File: docs/tutorials/onnx/inference_on_onnx_model.md [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+```python
+numpy_path = os.path.join(model_folder, current_model, 'test_data_0.npz')
+sample = np.load(numpy_path, encoding='bytes')
+inputs = sample['inputs']
+outputs = sample['outputs']
+```
+
+
+```python
+print("Input format: {}".format(inputs[0].shape))
```
Comment (safrooze) [inline docs/tutorials/onnx/inference_on_onnx_model.md]: What's the output format? Why do you use `outputs[0]` when you're using argmax during testing?

File: docs/tutorials/onnx/inference_on_onnx_model.md [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+```python
+TOP_N = 3 # How many top guesses we show in the visualization
+```
+
+
+Transform function to set the data into the format the network requires (N, 3, 224, 224)
+
+
+```python
+transform = lambda img: np.expand_dims(np.transpose(img, (2,0,1)),axis=0).astype(np.float32)
```
Comment (safrooze) [inline docs/tutorials/onnx/inference_on_onnx_model.md]: Recommend writing a function instead of using lambda (easier to read).

File: docs/tutorials/onnx/inference_on_onnx_model.md [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+## Test using real images
+
+
+```python
+TOP_N = 3 # How many top guesses we show in the visualization
+```
+
+
+Transform function to set the data into the format the network requires (N, 3, 224, 224)
```
Comment (safrooze) [inline docs/tutorials/onnx/inference_on_onnx_model.md]: Would be good to explain what N is here (batch-size), specially that right above you use N to refer to number of predictions to display. Might also be a good idea to rename `TOP_N` to `TOP_K` or something else.

File: docs/tutorials/onnx/inference_on_onnx_model.md [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+```python
+plot_predictions(caltech101_images, result[3:7], categories, TOP_N)
+```
+
+
+![png](https://github.com/ThomasDelteil/web-data/blob/c77c2e93ba142f45682ed63c191d2568b20aff25/mxnet/doc/tutorials/onnx/caltech101.png?raw=true)<!--no-notebook-->
+
+
+**Hmm, not so good...** We see where the network is coming from but effectively `wrench`, `dolphin` and `lotus` categories are not in the ImageNet classes...
```
Comment (safrooze) [inline docs/tutorials/onnx/inference_on_onnx_model.md]: Suggest rewording to something like this: "Even though predictions are close, they are not accurate, which is due to the fact that ImageNet dataset does not contain `wrench`, `dolphin`, or `lotus` categories and our network has been trained on ImageNet."

File: docs/tutorials/onnx/inference_on_onnx_model.md [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+Lucky for us, the [Caltech101 dataset](http://www.vision.caltech.edu/Image_Datasets/Caltech101/) has them, let's see if we can fine-tune our network to classify those correctly.
```
Comment (safrooze) [inline docs/tutorials/onnx/inference_on_onnx_model.md]: let's see **how** we can...classify **these categories** correctly.

File: docs/tutorials/onnx/inference_on_onnx_model.md [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+
+### Downloading supporting files
+These are images and a vizualisation script
+
+
+```python
+image_folder = "images"
+utils_file = "utils.py" # contain utils function to plot nice visualization
+images = ['apron', 'hammerheadshark', 'dog', 'wrench', 'dolphin', 'lotus']
+base_url = "https://github.com/ThomasDelteil/web-data/blob/c77c2e93ba142f45682ed63c191d2568b20aff25/mxnet/doc/tutorials/onnx/{}?raw=true"
```
[the remark this answers is not in the document]
Comment (ThomasDelteil) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: yes

File: docs/tutorials/onnx/inference_on_onnx_model.md [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+```
+
+
+```python
+from utils import *
+```
+
+## Downloading a model from the [ONNX model zoo](https://github.com/onnx/models)
+
+We download a pre-trained model, in our case the [vgg16](https://arxiv.org/abs/1409.1556) model, trained on [ImageNet](www.image-net.org/). The model comes packaged in an archive `tar.gz` file containing an `model.onnx` model file and some sample input/output data.
```
[the remark this answers is not in the document]
Comment (ThomasDelteil) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: thanks will update

File: docs/tutorials/onnx/inference_on_onnx_model.md [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+
+```python
+ctx = mx.cpu()
+```
+
+And load them into a MXNet Gluon symbol block
+
+
+```python
+net = gluon.nn.SymbolBlock(outputs=sym, inputs=mx.sym.var('input_0'))
```
[the remark this answers is not in the document]
Comment (ThomasDelteil) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: Sure, I will add that this is the default ONNX input data name

File: docs/tutorials/onnx/inference_on_onnx_model.md [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+net_params = net.collect_params()
+for param in params:
+    if param in net_params:
```
[the remark this answers is not in the document]
Comment (ThomasDelteil) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: Yes, it can happen for example when you are removing layers and fine-tuning. I leave it here so that it is easier to do a copy paste.

File: docs/tutorials/onnx/inference_on_onnx_model.md [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+        net_params[param]._load_init(params[param], ctx=ctx)
+```
+
+This [hybridizes](https://mxnet.incubator.apache.org/tutorials/gluon/hybrid.html) the network the computational graph gets cached and we get performance gains
```
[the remark this answers is not in the document]
Comment (ThomasDelteil) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: Yes I agree with that, just left it here so that it gets a bit of visibility and people think about hybridizing their gluon nets

File: docs/tutorials/onnx/inference_on_onnx_model.md (hunk 3 above)
Comment (ThomasDelteil) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: not sure what you mean? Do you mean I should show that the ouput is (1000,1) ?

File: docs/tutorials/onnx/inference_on_onnx_model.md (hunk 4 above)
Comment (ThomasDelteil) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: ok will do

File: docs/tutorials/onnx/inference_on_onnx_model.md (hunk 5 above)
Comment (ThomasDelteil) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: good suggestion

File: docs/tutorials/onnx/inference_on_onnx_model.md (hunk 6 above)
Comment (ThomasDelteil) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: :+1:

File: docs/tutorials/onnx/inference_on_onnx_model.md (hunk 7 above)
Comment (ThomasDelteil) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: :+1:

File: docs/tutorials/onnx/inference_on_onnx_model.md [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+if not os.path.isfile(utils_file):
+    wget.download(base_url.format(utils_file))                               
+```
+
+
+```python
+from utils import *
+```
+
+## Downloading a model from the [ONNX model zoo](https://github.com/onnx/models)
```
[the remark this answers is not in the document]
Comment (ThomasDelteil) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: ok will update that

File: docs/tutorials/onnx/inference_on_onnx_model.md [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+inputs = sample['inputs']
+outputs = sample['outputs']
+```
+
+
+```python
+print("Input format: {}".format(inputs[0].shape))
+```
+
+    Input format: (1, 3, 224, 224) <!--no-notebook-->
```
[the remark this answers is not in the document]
Comment (ThomasDelteil) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: It won't be shown

File: docs/tutorials/onnx/inference_on_onnx_model.md (hunk 1 above)
Comment (ThomasDelteil) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: thanks

File: docs/tutorials/onnx/inference_on_onnx_model.md (hunk 3 above)
Comment (safrooze) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: Yes, showing the shape helps. But more specifically, I'm referring to this line:
`print("Loaded model and sample output predict the same class: {}".format(np.argmax(result) == np.argmax(outputs[0])))`
Here you're using outputs[0] for all the results. What's the type of `outputs`? What are other indices?

File: docs/tutorials/onnx/inference_on_onnx_model.md (hunk 2 above)
Comment (ThomasDelteil) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: not sure 😄 I want the output to be in a (1000,N) shape, basically flattening through the batches regardless of the number of batches. So I go through each prediction, store them in an array, and then convert this array to a np.array

File: docs/tutorials/onnx/inference_on_onnx_model.md [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+
+```python
+image_folder = "images"
+utils_file = "utils.py" # contain utils function to plot nice visualization
+images = ['apron', 'hammerheadshark', 'dog', 'wrench', 'dolphin', 'lotus']
+base_url = "https://github.com/ThomasDelteil/web-data/blob/c77c2e93ba142f45682ed63c191d2568b20aff25/mxnet/doc/tutorials/onnx/{}?raw=true"
+
+if not os.path.isdir(image_folder):
+    os.makedirs(image_folder)
+    for image in images: wget.download(base_url.format("{}/{}.jpg".format(image_folder, image)), image_folder)
```
Comment (Ishitori) [inline docs/tutorials/onnx/inference_on_onnx_model.md]: I would write wget on a separate line to follow PEP8 standard

File: docs/tutorials/onnx/inference_on_onnx_model.md (hunk 2 above)
Comment (safrooze) [reply docs/tutorials/onnx/inference_on_onnx_model.md]: So you want to transpose from (N,1000) to (1000,N)?

**Comment by ThomasDelteil** [comment]
@safrooze @Ishitori see updated version

**Comment by ThomasDelteil** [comment]
@safrooze are you happy with the changes? I need to add this tutorial to my other MXNet PR to have it tested

**Comment by ThomasDelteil** [comment]
@safrooze ping
